src/main/java/gregtech/api/capability/impl/MultiblockRecipeLogic.java

```diff
@@ -71,13 +85,76 @@ protected IMultipleTankHandler getOutputTank() {
     @Override
     protected void trySearchNewRecipe() {
         // do not run recipes when there are more than 5 maintenance problems
+        // Maintenance can apply to all multiblocks, so cast to a base multiblock class
         MultiblockWithDisplayBase controller = (MultiblockWithDisplayBase) metaTileEntity;
-        if (controller.hasMaintenanceMechanics() && controller.getNumMaintenanceProblems() > 5)
+        if (controller.hasMaintenanceMechanics() && controller.getNumMaintenanceProblems() > 5) {
             return;
+        }
+
+        // Distinct buses only apply to some of the multiblocks, so check the controller against a lower class
+        if(controller instanceof RecipeMapMultiblockController) {
+            RecipeMapMultiblockController distinctController = (RecipeMapMultiblockController) controller;
+
+            if(distinctController.canBeDistinct() && distinctController.isDistinct()) {
+                trySearchNewRecipeDistinct();
+                return;
+            }
+        }
 
         super.trySearchNewRecipe();
     }
 
+    protected void trySearchNewRecipeDistinct() {
+        long maxVoltage = getMaxVoltage();
+        Recipe currentRecipe = null;
+        List<IItemHandlerModifiable> importInventory = getInputBuses();
+        IMultipleTankHandler importFluids = getInputTank();
+
+        // Our caching implementation
+        // This guarantees that if we get a recipe cache hit, our efficiency is no different from other machines
+        if (previousRecipe != null && previousRecipe.matches(false, importInventory.get(lastRecipeIndex), importFluids)) {
+            currentRecipe = previousRecipe;
+            // TODO, is setting the invalidInputs here needed?
+            invalidInputsForRecipes = false;
+            // If a valid recipe is found, immediately attempt to return it to prevent inventory scanning
+            if (setupAndConsumeRecipeInputs(currentRecipe, importInventory.get(lastRecipeIndex))) {
+                setupRecipe(currentRecipe);
+                metaTileEntity.getNotifiedItemInputList().clear();
+                metaTileEntity.getNotifiedFluidInputList().clear();
+                //TODO, should we set the previous recipe to the current recipe here, or would that cause some issues with caching once parallel is introduced?
```

Comment:
i think we should remove the current bus from the list, not clear it all

Reply:
Regarding Trouser's comment, I have done that.

Regarding dan's comment, I wanted to set the previous recipe to the current recipe here, as it is already done in `trySearchNewRecipe` for non-distinct buses. However, after thinking about it more, I believe I do not need to set the previous recipe to the current recipe here, because this bit of logic is where the previous recipe was not null and the current recipe matches, so the previous recipe is already the current recipe.

However, I do need to do this for a cache miss, in the lower bit of logic, since I want to cache the previous recipe. However, I am already doing this.